src/form/input/Input.tsx

```diff
@@ -109,21 +141,24 @@ function Input(props: InputProps) {
   function handleChange(event: React.SyntheticEvent<HTMLInputElement>) {
     if (isNumberInput) {
       const {value: newValue} = event.currentTarget;
+      let formattedNewValue = parseNumber(newValue, locale);
 
-      let formattedNewValue = newValue
-        .replace(NOT_INTEGER_FIRST_CHARACTER_OF_STRING_REGEX, "")
-        .replace(NOT_NUMBER_NOR_DECIMAL_POINT_REGEX, "")
-        .replace(PRECISION_REGEX, "$1");
-
-      if (maxFractionDigits > 0) {
-        const decimalNumberParts = newValue.split(DECIMAL_NUMBER_SEPARATOR);
+      if (
+        IS_LAST_CHARACTER_DECIMAL_POINT_REGEX.test(formattedNewValue) ||
+        MATCH_ZEROS_AFTER_DECIMAL_REGEX.test(formattedNewValue)
+      ) {
+        const decimalNumberParts = formattedNewValue.split(DECIMAL_NUMBER_SEPARATOR);
         const decimalPart = decimalNumberParts[1];
-        const integerPart = decimalNumberParts[0].replace(THOUSANDTHS_SEPARATOR, "");
+        const integerPart = decimalNumberParts[0];
+
+        formattedNewValue = `${integerPart}${DECIMAL_NUMBER_SEPARATOR}${decimalPart}`;
+      }
```

Comment:
I don't think we require this part any longer

Reply:
When we remove this condition, the user can enter any character. I guess we need this condition both for formatting and parsing. Maybe we can create a new utility for this. What do you think?

Reply:
Hmm. Not sure I understand. Doesn't this just check if the entered value is a fraction and then re-construct the value in natural number format ? What characters are we trying to prevent users to use with this if block ?

Reply:
I think users can enter only number and decimal operator.

Reply:
How do we check against the decimal operator ?

Reply:
Nevermind. We are testing against `formattedNewValue` which is in scientific format so `IS_LAST_CHARACTER_DECIMAL_POINT_REGEX` and `MATCH_ZEROS_AFTER_DECIMAL_REGEX` works properly here.

Reply:
> When we remove this condition, the user can enter any character

What does `parseNumber` return if the event.currentTarget.value is something like "10a" ?

Reply:
`event.currentTarget.value` is returning `"10a"` in this version. I think we should prevent entering other characters without number or decimal separator in the handler.

I will think how we can combine your above suggestion with this.